tests/test_job.py

```diff
@@ -131,6 +131,39 @@ def test_str(self):
         job = self.project.open_job({"a": 0})
         assert str(job) == job.id
 
+    def test_eq(self):
+        job = self.project.open_job({"a": 0})
+
+        # Make sure that Jobs can only be equal to other Job instances.
+        class NonJob:
+            """Minimal class that cannot be compared with Job objects."""
+
+            def __init__(self, job):
+                self.id = job.id
+                self._workspace = job.workspace()
+
+        class JobSubclass(Job):
+            """Minimal subclass that can be compared with Job objects."""
+
+            def __init__(self, job):
+                self._id = job.id
+                self._workspace = job.workspace()
+
+            def workspace(self):
+                return self._workspace
+
+        non_job = NonJob(job)
+        assert job != non_job
+        assert non_job != job
```

Comment:
Why does this ensure that we have `NotImplemeted` as opposed to `False`?

Reply:
I'm testing both `a != b` and `b != a` because of the order in which Python resolves `__eq__` operators. This section of the Python docs explains why both tests are relevant here, and why we return `NotImplemented` here: https://docs.python.org/3/reference/datamodel.html#object.__eq__

Specifically, the order of resolution is controlled by this rule: "If the operands are of different types, and right operand’s type is a direct or indirect subclass of the left operand’s type, the reflected method of the right operand has priority, otherwise the left operand’s method has priority."

Reply:
Also, I used this issue as a reference implementation because it appeared to correct the equality behavior that I was unsure about for subclasses (`JobSubclass`) and objects of different types with similar APIs (`NonJob`): https://github.com/googleapis/google-cloud-python/issues/3455